### PR TITLE
Move to parsing from_json input preserving quoted strings.

### PR DESCRIPTION
**Pull request by revans2**
This is based on some of the great work that @andygrove did https://github.com/NVIDIA/spark-rapids/pull/10326 but has evolved quite a bit past that.

With this we are now asking CUDF to return all leaf nodes as strings and to preserve the quotes on those strings if they are there.  This lets us handle parsing different data types in a spark compatible way.  For example when parsing a decimal value from the following data a, b, and  d are valid decimal input, but c is not.
```
{"a":12345}
{"b":"12345"}
{"c":12,345}
{"d":"12,345"}
```

This does not fix all of those use cases. It just lays the ground work to be able to fix them, ideally we would fix this with a small number of custom kernels after we have made the code common between from_json/JsonToStructs and ScanJson.

Because the code is not common we cannot close some of the issues that this fixes because the issues were filed against both JsonToStructs and ScanJson.

This has a dependency on something that was just merged into CUDF today so it will fail to compile, but I would love to have some reviews for this before then.

There are two small regressions introduced by this. We no longer support duplicate struct keys and fall back to the CPU if we see them. We could make it work by injecting nulls for one of the keys, but it should be a really rare case, and it totally messes up SQL/dataframe access to those columns so I don't think it is a priority.

I also have not finished fixing up date/timestamp parsing for unquoted numbers. The old code worked only when all of the values in the batch were that type, but didn't work if they were mixed.

Performance Results:

In my testing I saw about a 4% slow down for simple queries compared to the current code. I would need to do some traces to see how much of that we can expect to get back with custom kernels for conversion and such, but I would expect that we can speed it up significantly over time.

### CODE REVIEW

**Comment by revans2** [comment]
UPDATE: I pushed another performance improvement that gets the performance on par with the original code. I still think that there is more that we can do to speed it up, but this should remove any concerns about this particular PRs performance.

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -3663,10 +3663,28 @@ object GpuOverrides extends Logging {
...
+            val fn = st.fieldNames
+            if (fn.length != fn.distinct.length) {
+              true
+            } else {
+              st.fields.exists(f => hasDuplicateFieldNames(f.dataType))
+            }
+          case ArrayType(ct, _) => hasDuplicateFieldNames(ct)
+          case MapType(kt, vt, _) => hasDuplicateFieldNames(kt) || hasDuplicateFieldNames(vt)
+          case _ => false
+        }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3677]:
```suggestion
        def hasDuplicateFieldNames(dt: DataType): Boolean = {
          TrampolineUtil.dataTypeExistsRecursively(dt,
            {
              case st: StructType =>
                val fn = st.fieldNames
                fn.length != fn.distinct.length
              case _ => false
            })
        }
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 2]
```diff
@@ -66,6 +66,150 @@ class JsonDeviceDataSource(combined: ColumnVector) extends DataSource {
...
+      populateSchema(at.elementType, "element", child)
+    case st: StructType =>
+      val child = builder.addColumn(DType.STRUCT, name)
+      for (sf <- st.fields) {
+        populateSchema(sf.dataType, sf.name, child)
+      }
+    case _: MapType =>
+      throw new IllegalArgumentException("MapType is not supported yet for schema conversion")
+    case _ =>
+      // TODO do we need to read all of the leaf nodes as strings???
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:83]: followup needed?

**Review by jlowe — DISMISSED** [review]
small nits, lgtm.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
@jlowe please take another look

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**